### ISSUE TITLE
Shutdown race

### ISSUE THREAD

**Issue by matheusd** [decred/dcrd#2037]
```2020-01-22 16:05:08.285 [INF] DCRD: Received signal (interrupt).  Shutting down...
2020-01-22 16:05:08.285 [WRN] SRVR: Server shutting down
2020-01-22 16:05:08.285 [WRN] RPCS: RPC server shutting down
2020-01-22 16:05:08.291 [INF] RPCS: RPC server shutdown complete
2020-01-22 16:05:08.292 [INF] BMGR: Block manager shutting down
2020-01-22 16:05:08.294 [INF] SRVR: seeder 'testnet-seed-1.decred.org' error: Get https://testnet-seed-1.decred.org/api/addrs?ipversion=0&pver=0&services=1: context canceled
==================
WARNING: DATA RACE
Write at 0x00c00ec2a6a8 by goroutine 33:
  sync.(*WaitGroup).Wait()
      /home/user/apps/go/src/internal/race/race.go:41 +0xee
  main.(*rpcServer).Stop()
      /home/user/projetos/decred/dcrd/rpcwebsocket.go:1294 +0x1b7
  main.(*server).Run()
      /home/user/projetos/decred/dcrd/server.go:2553 +0x3cd
  main.dcrdMain.func5()
      /home/user/projetos/decred/dcrd/dcrd.go:194 +0x4c

Previous read at 0x00c00ec2a6a8 by goroutine 36:
  sync.(*WaitGroup).Add()
      /home/user/apps/go/src/internal/race/race.go:37 +0x18d
  main.(*wsNotificationManager).Start()
      /home/user/projetos/decred/dcrd/rpcwebsocket.go:1286 +0x4b
  main.(*rpcServer).Start()
      /home/user/projetos/decred/dcrd/rpcserver.go:5499 +0x42a
  main.(*server).Run.func1()
      /home/user/projetos/decred/dcrd/server.go:2514 +0x59
```

I believe this is due to the `wsNotificationManager.Start()` being called from a different goroutine than `wsNotificationManager.Stop()` (starting [here](https://github.com/decred/dcrd/blob/e3547d0822db291479ea67a914feaa70806f0828/server.go#L2513-L2514)).

A way to solve this would be to have all subsystems start their own handling goroutines (instead of having `server.Run()` do it) while the server itself calls each `Start()` serially. 

This is a low priority race since it involves node shutdown and can't really cause data loss. I'm creating the issue just to have it documented.

**Comment by jrick** [comment]
https://godoc.org/sync#WaitGroup.Add

> Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.

This violates the Go Memory Model and needs a fix. There is no such thing as a benign data race.

**Comment by jrick** [comment]
Minimal repro:

```
$ cat /tmp/wg.go
package main

import "sync"

func main() {
	var wg sync.WaitGroup
	for {
		go func() {
			wg.Add(1)
			wg.Done()
		}()
		wg.Wait()
	}
}
$ go run -race /tmp/wg.go
==================
WARNING: DATA RACE
Write at 0x00c000018078 by main goroutine:
  sync.(*WaitGroup).Wait()
      /usr/local/go/src/internal/race/race.go:41 +0xee
  main.main()
      /tmp/wg.go:12 +0x99

Previous read at 0x00c000018078 by goroutine 28:
  sync.(*WaitGroup).Add()
      /usr/local/go/src/internal/race/race.go:37 +0x18d
  main.main.func1()
      /tmp/wg.go:9 +0x41

Goroutine 28 (finished) created at:
  main.main()
      /tmp/wg.go:8 +0x8b
==================
==================
WARNING: DATA RACE
Read at 0x00c000018078 by goroutine 29:
  sync.(*WaitGroup).Add()
      /usr/local/go/src/internal/race/race.go:37 +0x18d
  main.main.func1()
      /tmp/wg.go:9 +0x41

Previous write at 0x00c000018078 by main goroutine:
  sync.(*WaitGroup).Wait()
      /usr/local/go/src/internal/race/race.go:41 +0xee
  main.main()
      /tmp/wg.go:12 +0x99

Goroutine 29 (running) created at:
  main.main()
      /tmp/wg.go:8 +0x8b
==================
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 1 [running]:
sync.(*WaitGroup).Wait(0xc000018070)
	/usr/local/go/src/sync/waitgroup.go:132 +0x142
main.main()
	/tmp/wg.go:12 +0x9a
exit status 2
```

**Comment by dnldd** [comment]
On it.
